source/app/controllers/urls_controller.rb

```diff
@@ -1,2 +1,38 @@
 class UrlsController < ApplicationController
+  before_action :set_url, only: [:show]
+
+  def index
+    @urls = Url.all
+  end
+
+  def new
+    @url = Url.new
+  end
+
+  def create
+    @url = Url.find_or_initialize_by(url_params)
+    if @url.save
+      redirect_to urls_path
+    else
+      flash[:error] = @url.errors.full_messages
+      redirect_to new_url_path(@url)
+    end
+  end
+
+  def show
+    @url.click_count += 1
+    @url.save
+
+    redirect_to "#{@url.long_url}"
+  end
+
+  private
+
+    def set_url
+      @url = Url.find_by(short_url: params[:short_url])
+    end
+
+    def url_params
+      params.require(:url).permit(:long_url, :short_url)
```

Comment:
We permit setting the `short_url` here, but there isn't an input in the form for it, and it will be overwritten in the model. I think `short_url` should be removed here.